trieste/models/optimizer.py

```diff
@@ -0,0 +1,134 @@
+from typing import Any, Dict, Optional, Callable, Union, Iterable, Tuple
+from functools import singledispatch
+from dataclasses import dataclass, field
+import tensorflow as tf
+import gpflow
+import scipy
+
+from gpflow.models import ExternalDataTrainingLossMixin, InternalDataTrainingLossMixin
+from ..utils import jit
+from ..data import Dataset
+
+
+DatasetTransformer = Callable[[Dataset, Optional[int]], Union[Iterable, Tuple]]
+"""
+Type alias for a function that maps a dataset from a :class:`~optiflow.datasets.Dataset`
+and generates an iterator over the transformation.
+"""
+
+LossClosure = Callable[[], tf.Tensor]
+"""
+Type alias for a loss closure that is used generally for optimizing.
+"""
+
+OptimizeResult = Union[scipy.optimize.OptimizeResult, None]
```

Comment:
is this name going to be confusing, as we already have an `OptimizationResult` for the main optimizer?

Reply:
Should we rename the main optimizer result?

Reply:
what would you suggest to? I think it's currently quite a good name. Indeed, I was hoping to rename it to `OptimizeResult` cos it's shorter. You've only used this alias once. Is it indispensable?

Reply:
`OptimizeResult` sounds like an action. `OptimisationResult` or `OptimizerResult` make more sense